cmds/tail/tail.go

```diff
@@ -0,0 +1,259 @@
+}
+
+// getBlockSize returns the number of bytes to read for each ReadAt call. This
+// helps minimize the number of syscalls to get the last N lines of the file.
+func getBlockSize(numLines int) int64 {
+	// This is currently computed as 81 * N, where N is the requested number of
+	// lines, and 81 is a relatively generous estimation of the average line
+	// length.
+	return 81 * int64(numLines)
+}
+
+// readLastLinesBackwards reads the last N lines from the provided file, reading
+// backwards from the end of the file. This is more efficient than reading from
+// the beginning, but can only be done on seekable files, (e.g. this won't work
+// on stdin). For non-seekable files see readLastLinesFromBeginning.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesBackwards(inFile *os.File, writer io.Writer, numLines int) error {
+	blkSize := getBlockSize(numLines)
+	// go to the end of the file
+	lastPos, err := inFile.Seek(0, os.SEEK_END)
+	if err != nil {
+		return err
+	}
+	// read block by block backwards until `numLines` lines are found
+	readData := make([]byte, 0)
+	buf := make([]byte, blkSize)
+	pos := lastPos
+	foundLines := 0
+	// for each block, count how many new lines, until they add up to `numLines`
+	for {
+		if pos == 0 {
+			break
+		}
+		var thisChunkSize int64
+		if pos < blkSize {
+			thisChunkSize = pos
+		} else {
+			thisChunkSize = blkSize
+		}
+		pos -= thisChunkSize
+		n, err := inFile.ReadAt(buf, pos)
+		if err != nil {
+			if err == io.EOF {
+				// if we got EOF but we also received some data, go ahead and handle
+				// what we got
+				if n == 0 {
+					break
+				}
+			} else {
+				return err
+			}
+		}
+		// merge this block to what was read so far
+		readData = append(buf[:n], readData...)
+		// count how many lines we have so far, and stop reading if we have
+		// enough
+		foundLines += bytes.Count(buf[:n], []byte{'\n'})
+		if foundLines >= numLines {
+			break
+		}
+	}
+	// find the start of the n-th to last line
+	slice := readData
+	// `data` contains up to `numLines` lines of the file
+	var data []byte
+	if len(slice) != 0 {
+		if slice[len(slice)-1] == '\n' {
+			// don't consider the last new line for the line count
+			slice = slice[:len(slice)-1]
+		}
+		var (
+			foundLines = 0
+			idx        int
+		)
+		for {
+			if foundLines >= numLines {
+				break
+			}
+			// find newlines backwards from the end of `slice`
+			idx = bytes.LastIndexByte(slice, '\n')
+			if idx == -1 {
+				// there are less than `numLines` lines
+				break
+			}
+			foundLines++
+			slice = slice[:idx-1]
+		}
+		if idx == -1 {
+			// if there are less than `numLines` lines, use all what we have read
+			data = readData
+		} else {
+			data = readData[idx+1:] // +1 to skip the newline belonging to the previous line
+		}
+	}
+	// write the requested lines to the writer
+	if _, err = writer.Write(data); err != nil {
```

Comment:
if you do move the seek out, do this:

```go
_, err = writer.Write(data)
return err
```
  
or if you keep the seek, do it to the call below.

Reply:
ack. For now I've done it in the call below. I'll wait for the discussion at the comment above to eventually do it above, https://github.com/u-root/u-root/pull/548#discussion_r159772382